executor/brie.go

```diff
@@ -0,0 +1,462 @@
+		b.err = errors.Annotate(err, "invalid destination URL")
+		return nil
+	}
+
+	switch storageURL.Scheme {
+	case "s3":
+		storage.ExtractQueryParameters(storageURL, &cfg.S3)
+	case "gs", "gcs":
+		storage.ExtractQueryParameters(storageURL, &cfg.GCS)
+	default:
+		break
+	}
+
+	cfg.Storage = storageURL.String()
+	e.info.storage = cfg.Storage
+
+	for _, opt := range s.Options {
+		switch opt.Tp {
+		case ast.BRIEOptionRateLimit:
+			cfg.RateLimit = opt.UintValue
+		case ast.BRIEOptionConcurrency:
+			cfg.Concurrency = uint32(opt.UintValue)
+		case ast.BRIEOptionChecksum:
+			cfg.Checksum = opt.UintValue != 0
+		case ast.BRIEOptionSendCreds:
+			cfg.SendCreds = opt.UintValue != 0
+		}
+	}
+
+	if len(s.Tables) != 0 {
+		cfg.Filter.DoTables = make([]*filter.Table, 0, len(s.Tables))
+		for _, tbl := range s.Tables {
+			cfg.Filter.DoTables = append(cfg.Filter.DoTables, &filter.Table{
+				Name:   tbl.Name.O,
+				Schema: tbl.Schema.O,
+			})
+		}
+	}
+
+	switch s.Kind {
+	case ast.BRIEKindBackup:
+		e.backupCfg = &task.BackupConfig{Config: cfg}
+
+		for _, opt := range s.Options {
+			switch opt.Tp {
+			case ast.BRIEOptionLastBackupTS:
+				tso, err := b.parseTSString(opt.StrValue)
+				if err != nil {
+					b.err = err
+					return nil
+				}
+				e.backupCfg.LastBackupTS = tso
+			case ast.BRIEOptionLastBackupTSO:
+				e.backupCfg.LastBackupTS = opt.UintValue
+			case ast.BRIEOptionBackupTimeAgo:
+				e.backupCfg.TimeAgo = time.Duration(opt.UintValue)
+			case ast.BRIEOptionBackupTSO:
+				e.backupCfg.BackupTS = opt.UintValue
+			case ast.BRIEOptionBackupTS:
+				tso, err := b.parseTSString(opt.StrValue)
+				if err != nil {
+					b.err = err
+					return nil
+				}
+				e.backupCfg.BackupTS = tso
+			}
+		}
+
+	case ast.BRIEKindRestore:
+		e.restoreCfg = &task.RestoreConfig{Config: cfg}
+		for _, opt := range s.Options {
+			switch opt.Tp {
+			case ast.BRIEOptionOnline:
+				e.restoreCfg.Online = opt.UintValue != 0
+			}
+		}
+
+	default:
+		b.err = errors.Errorf("unsupported BRIE statement kind: %s", s.Kind)
+		return nil
+	}
+
+	return e
+}
+
+// BRIEExec represents an executor for BRIE statements (BACKUP, RESTORE, etc)
+type BRIEExec struct {
+	baseExecutor
+
+	backupCfg  *task.BackupConfig
+	restoreCfg *task.RestoreConfig
+	info       *brieTaskInfo
+}
+
+// Next implements the Executor Next interface.
+func (e *BRIEExec) Next(ctx context.Context, req *chunk.Chunk) error {
+	req.Reset()
+	if e.info == nil {
+		return nil
+	}
+
+	bq := globalBRIEQueue
+
+	e.info.connID = e.ctx.GetSessionVars().ConnectionID
+	e.info.queueTime = types.CurrentTime(mysql.TypeDatetime)
+	taskCtx, taskID := bq.registerTask(ctx, e.info)
+	defer bq.cancelTask(taskID)
+
+	progress, err := bq.acquireTask(taskCtx, taskID)
+	if err != nil {
+		return err
+	}
+	defer bq.releaseTask()
+
+	e.info.execTime = types.CurrentTime(mysql.TypeDatetime)
+	glue := &tidbGlueSession{se: e.ctx, progress: progress, info: e.info}
+
+	// manually monitor the Killed status...
+	go func() {
```

Comment:
Should we move this block to line 308, or we'll wait on workerCh forever even if kill is executed?